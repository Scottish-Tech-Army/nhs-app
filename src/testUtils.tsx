import React from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "./data/store";
import boxContentsReducer, { initialState } from "./data/BoxContentsSlice";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProvider(
  component: React.ReactElement,
  {
    preloadedState = { boxContents: initialState },
    store = configureStore({
      reducer: { boxContents: boxContentsReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {

  return {
    store,
    ...render(<Provider store={store}>{component}</Provider>, renderOptions),
  };
}
