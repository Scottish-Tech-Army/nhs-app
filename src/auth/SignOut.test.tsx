import React from "react";
import { SignOut } from "./SignOut";
import { signOut } from "../model/auth/AuthActions";
import { screen, waitFor } from "@testing-library/react";
import { renderWithProvider } from "../testUtils";

jest.mock("../model/auth/AuthActions");

describe("component SignOut", () => {
  beforeEach(() => {
    (signOut as jest.Mock).mockImplementation(() => () => "dummy action");
  });

  it("logout popup confirm logs user out", async () => {
    const { user, container } = renderWithProvider(<SignOut />);

    await user.click(screen.getByLabelText("log out"));

    expect(screen.getByText("Are you sure you want to leave?")).toBeDefined();
    expect(container).toMatchSnapshot();

    await user.click(screen.getByRole("button", { name: "YES" }));

    expect(signOut).toHaveBeenCalledTimes(1);
  });

  it("logout popup cancel does not log user out", async () => {
    const { user } = renderWithProvider(<SignOut />);

    await user.click(screen.getByLabelText("log out"));

    expect(screen.getByText("Are you sure you want to leave?")).toBeDefined();

    await user.click(screen.getByRole("button", { name: "Cancel" }));

    await waitFor(() =>
      expect(screen.queryByText("Are you sure you want to leave?")).toBeNull()
    );

    expect(signOut).not.toHaveBeenCalled();
  });
});
