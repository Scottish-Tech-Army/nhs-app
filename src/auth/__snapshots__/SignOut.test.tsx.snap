// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component SignOut logout popup confirm logs user out 1`] = `
<div>
  <div
    aria-label="log out"
    class="navicon"
    id="auth-signout"
  >
    <svg>
      logout.svg
    </svg>
    <div>
      Log Out
    </div>
  </div>
  <div
    class="confirm-reset-popup"
  >
    <h2>
      Log out
    </h2>
    <div
      class="message"
    >
      Are you sure you want to leave?
    </div>
    <button
      class="confirm-yes"
      type="button"
    >
      YES
    </button>
    <button
      class="confirm-no"
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    class="modal-overlay"
  />
</div>
`;
